src/java.base/share/classes/sun/security/rsa/RSASignature.java

```diff
@@ -213,10 +213,12 @@ protected boolean engineVerify(byte[] sigBytes) throws SignatureException {
                     RSACore.getByteLength(publicKey));
             }
             byte[] digest = getDigestValue();
+            // https://www.rfc-editor.org/rfc/rfc8017.html#section-8.2.2
+            // Step 4: Compare the encoded message, not the decoded.
+            byte[] encoded = RSAUtil.encodeSignature(digestOID, digest);
             byte[] decrypted = RSACore.rsa(sigBytes, publicKey);
             byte[] unpadded = padding.unpad(decrypted);
```

Comment:
It looks like safer to me that decodeSignature() method is removed and the same method get used for the verification.

Maybe, these two lines could be moved ahead (at line 215) so that the processing order is consistent with the RFC 8017 described order, for readability.